packages/next/next-server/server/image-optimizer.ts

```diff
@@ -0,0 +1,240 @@
+    res.end('"url" parameter is not allowed')
+    return { finished: true }
+  }
+
+  if (!w) {
+    res.statusCode = 400
+    res.end('"w" parameter (width) is required')
+    return { finished: true }
+  } else if (Array.isArray(w)) {
+    res.statusCode = 400
+    res.end('"w" parameter (width) cannot be an array')
+    return { finished: true }
+  }
+
+  if (!q) {
+    res.statusCode = 400
+    res.end('"q" parameter (quality) is required')
+    return { finished: true }
+  } else if (Array.isArray(q)) {
+    res.statusCode = 400
+    res.end('"q" parameter (quality) cannot be an array')
+    return { finished: true }
+  }
+
+  const width = parseInt(w, 10)
+
+  if (!width || isNaN(width)) {
+    res.statusCode = 400
+    res.end('"w" parameter (width) must be a number greater than 0')
+    return { finished: true }
+  }
+
+  if (!sizes.includes(width)) {
+    res.statusCode = 400
+    res.end(`"w" parameter (width) of ${width} is not allowed`)
+    return { finished: true }
+  }
+
+  const quality = parseInt(q)
+
+  if (isNaN(quality) || quality < 1 || quality > 100) {
+    res.statusCode = 400
+    res.end('"q" parameter (quality) must be a number between 1 and 100')
+    return { finished: true }
+  }
+
+  const { href } = absoluteUrl
+  const hash = getHash([CACHE_VERSION, href, width, quality, mimeType])
+  const imagesDir = join(distDir, 'cache', 'images')
+  const hashDir = join(imagesDir, hash)
+  const now = Date.now()
+
+  if (await fileExists(hashDir, 'directory')) {
+    const files = await promises.readdir(hashDir)
+    for (let file of files) {
+      const [filename, extension] = file.split('.')
+      const expireAt = Number(filename)
+      const contentType = getContentType(extension)
+      if (now < expireAt) {
+        if (contentType) {
+          res.setHeader('Content-Type', contentType)
+        }
+        createReadStream(join(hashDir, file)).pipe(res)
+        return { finished: true }
+      } else {
+        await promises.unlink(join(hashDir, file))
+      }
+    }
+  }
+
+  const upstreamRes = await fetch(href)
+
+  if (!upstreamRes.ok) {
+    server.logError(
+      new Error(`Unexpected status ${upstreamRes.status} from upstream ${href}`)
+    )
+    return { finished: true }
+  }
+
+  const upstreamBuffer = Buffer.from(await upstreamRes.arrayBuffer())
+  const upstreamType = upstreamRes.headers.get('Content-Type')
+  const maxAge = getMaxAge(upstreamRes.headers.get('Cache-Control'))
+  const expireAt = maxAge * 1000 + now
+  let contentType: string
+
+  if (mimeType) {
+    contentType = mimeType
+  } else if (upstreamType?.startsWith('image/') && getExtension(upstreamType)) {
+    contentType = upstreamType
+  } else {
+    contentType = JPEG
+  }
+
+  if (!sharp) {
+    try {
+      // eslint-disable-next-line import/no-extraneous-dependencies
+      sharp = require('sharp')
+    } catch (error) {
+      error.message +=
+        '\nDid you forget to add it to "dependencies" in `package.json`?'
```

Comment:
```suggestion
      console.error(
        "To use Next.js' built-in image optimizer, you first need to install `sharp`.\n" +
        "Run `npm i sass` or `yarn add sass` inside your workspace.\n"
      )
```
Let's make this message similar to our [missing `sass`](https://github.com/vercel/next.js/blob/6ec36595e3687afd64b6aa43ca0ea2eb2d8380dc/packages/next/client/dev/error-overlay/format-webpack-messages.js#L105-L108) one, it might be good to add an err.sh similar to [this one](https://github.com/vercel/next.js/blob/canary/errors/install-sass.md) for it too

Reply:
I updated the message and added a markdown file to link to 👍